### ISSUE TITLE
Importing grpc._interceptor before mprofile causes SIGABRT with `free(): invalid pointer`

### ISSUE THREAD

**Issue by evgenii-petrov-arrival** [grpc/grpc#22977]
### What version of gRPC and what language are you using?
1.28.1 on Python 3.7.5

### What operating system (Linux, Windows,...) and version?
Linux, Ubuntu 18.04

### What runtime / compiler are you using (e.g. python version or version of gcc)
N/A , due to the use of wheels, I think, please correct me if I'm wrong here.

### What did you do?

```
$ python3.7 -m venv /tmp/mprofile-repro
$ source /tmp/mprofile-repro/bin/activate
$ python --version
Python 3.7.5

$ pip install --upgrade pip
Cache entry deserialization failed, entry ignored
Collecting pip
  Using cached https://files.pythonhosted.org/packages/54/2e/df11ea7e23e7e761d484ed3740285a34e38548cf2bad2bed3dd5768ec8b9/pip-20.1-py2.py3-none-any.whl
Installing collected packages: pip
  Found existing installation: pip 9.0.1
    Uninstalling pip-9.0.1:
      Successfully uninstalled pip-9.0.1
Successfully installed pip-20.1

$ pip install grpcio==1.28.1
Collecting grpcio==1.28.1
  Using cached grpcio-1.28.1-cp37-cp37m-manylinux2010_x86_64.whl (2.8 MB)
Collecting six>=1.5.2
  Using cached six-1.14.0-py2.py3-none-any.whl (10 kB)
Installing collected packages: six, grpcio
Successfully installed grpcio-1.28.1 six-1.14.0

$ pip install mprofile==0.0.10
Collecting mprofile==0.0.10
  Using cached mprofile-0.0.10-cp37-cp37m-manylinux2010_x86_64.whl (555 kB)
Installing collected packages: mprofile
Successfully installed mprofile-0.0.10

$ cat <<EOF > /tmp/mprofile-repro.py
import grpc._interceptor
import mprofile
EOF

$ python /tmp/mprofile-repro.py
free(): invalid pointer
Aborted

$ cat <<EOF > /tmp/mprofile-no-repro.py
import mprofile
import grpc._interceptor
EOF

$ python /tmp/mprofile-no-repro.py # success
```

### What did you expect to see?
I expected to not see SIGABRT . This is what happens when `mprofile` is imported before `grpc._interceptor`.

### What did you see instead?
SIGABRT with `free(): invalid pointer`.

### Anything else we should know about your project / environment?
A colleague on Mac OS can not reproduce this. Will try to post additional info about their setup later. Updated: Catalina 10.15.4 , python 3.7.6 from NixOS. Note that same NixOS setup on Linux, Ubuntu 18.04, also with python 3.7.6 fails, so python minor version is likely irrelevant here.

**Comment by lidizheng** [comment]
@evgenii-petrov-arrival Thank you for reporting this issue. Do you think you can help us further analysis this issue? 1) Can you check if the problem exist with `v1.29.0`? 2) Can you use `gdb` (any debugger) to trace what caused the failure?

**Comment by evgenii-petrov-arrival** [comment]
>Do you think you can help us further analysis this issue?
I'd be happy to.

>1) Can you check if the problem exist with v1.29.0?

Checked, it does.

>2) Can you use gdb (any debugger) to trace what caused the failure?

I've posted a log of running `backtrace` in `gdb -ex r --args python /tmp/mprofile-repro.py` to [this gist](https://gist.github.com/evgenii-petrov-arrival/3febfc769144ab65031adb3d2d582885).

Note that I can not reproduce this using `python3.7-dbg` on Ubuntu 18.04, but I'm not sure if mprofile or grpcio can be used with this build, due to `Third party extensions need a separate build to be used by this interpreter.` warning on the `python3.7-dbg` package.

I'd appreciate some guidance on what specific things in gdb to do, or which python builds to use.

**Comment by lidizheng** [comment]
Thanks for the tracing log. It shows that the SIGABRT was sent by `mprofiler` package. I can't yet find which pointer caused the "free(): invalid pointer" error. Normal gdb doesn't directly display Python frame information, you can use follow [DebuggingWithGdb](https://wiki.python.org/moin/DebuggingWithGdb) to setup gdb with Python debugging extensions. Can you try to find the culprit pointer?

**Comment by evgenii-petrov-arrival** [comment]
>you can use follow DebuggingWithGdb to setup gdb with Python debugging extensions

DebuggingWithGdb you've linked suggests using python-dbg package on Ubuntu, and I can't reproduce this issue with python from that package.

I'll try and see if I can reproduce this with a non-debug/debug builds created by `pyenv` and will report back with what I find.

In case that fails, any additional suggestions on how to move forward would be welcome.

**Comment by evgenii-petrov-arrival** [comment]
I've got pyenv working with following results:

`pyenv install 3.7.5` produces a build of python that has the same issue.

`pyenv install --debug 3.7.5` produces a build of python that does not have the same issue. Same build but with gcc flag `-O3`, instead of the usual `-O0` for the debug build also doesn't have this issue.

I tried producing debug symbols for the mprofile package, and ended up discovering that local build of mprofile, instead of the one uploaded to pypi, with no modifications, doesn't produce this error.

I no longer think that there is much point in trying to figure out why import ordering causes this invalid free, so I'm going to close this issue.
